libs/common/api_adapter.h

```diff
@@ -136,8 +138,43 @@ class ArnoldAPIAdapter {
     // Ideally GetPrimvars shouldn't be here
     virtual const std::vector<UsdGeomPrimvar> &GetPrimvars() const = 0;
 
+    const AtNodeEntry * GetCachedMtlxNodeEntry(const std::string &nodeEntryKey, const char *nodeDefinition, AtParamValueMap *params) {
+        // First we check if the nodeType is an arnold shader
+        std::lock_guard<AtMutex> lock(_nodeEntrymutex);
+        const auto shaderNodeEntryIt = _shaderNodeEntryCache.find(nodeEntryKey);
+        if (shaderNodeEntryIt == _shaderNodeEntryCache.end()) {
+            // NOTE for the future: we are in lock and the following function calls the system and query the disk
+            // This might be the source of contention or deadlock
+            const AtNodeEntry* nodeEntry = AiMaterialxGetNodeEntryFromDefinition(nodeDefinition, params);
+            _shaderNodeEntryCache[nodeEntryKey] = nodeEntry;
+            return nodeEntry;
+        }
+        return shaderNodeEntryIt->second;
+    };
+
+    AtString GetCachedOslCode(const std::string &oslCodeKey, const char *nodeDefinition, AtParamValueMap *params) {
+        std::lock_guard<AtMutex> lock(_oslCodeCacheMutex);
+        const auto oslCodeIt = _oslCodeCache.find(oslCodeKey);
+        if (oslCodeIt == _oslCodeCache.end()) {
+    #if ARNOLD_VERSION_NUM > 70104
+            _oslCodeCache[oslCodeKey] = AiMaterialxGetOslShaderCode(nodeDefinition, "shader", params);
+    #elif ARNOLD_VERSION_NUM >= 70104
```

Comment:
Is this the correct version number being compared against? It looks odd that it's comparing to the same value - basically this else statement says after the `if` says `ARNOLD_VERSION_NUM == 70104` right?

It also means that if the version would be `< 70104` this `if` statement body would effectively be empty?

Reply:
Yes, before Arnold 7.1.4 this API didn't exist and there was no way to get any osl code, so it won't render this shader.
In 7.1.4 the API was first introduced, and in following versions an argument was added. So that's why we have the current ifdefs. Indeed it would have behaved the same way if it had been written `elif ARNOLD_VERSION_NUM == 70104`

Reply:
Thanks, makes sense. 👍